pkg/pgmodel/querier/query_builder.go

```diff
@@ -216,68 +218,107 @@ type aggregators struct {
 	tsSeries    TimestampSeries //can be NULL and only present if timeClause == ""
 }
 
-/* The path is the list of ancestors (direct parent last) returned node is the most-ancestral node processed by the pushdown */
-// todo: investigate if query hints can have only node and lookback
-func getAggregators(md *promqlMetadata) (*aggregators, parser.Node, error) {
-	path := md.path // PromQL AST.
-	qh := md.queryHints
-	hints := md.selectHints
-	if !extension.ExtensionIsInstalled || qh == nil || hasSubquery(path) || hints == nil {
-		return getDefaultAggregators(), nil, nil
+// getAggregators returns the aggregator which should be used to fetch data for
+// a single metric. It may apply pushdowns to functions.
+func getAggregators(metadata *promqlMetadata) (*aggregators, parser.Node, error) {
+	// todo: investigate if query hints can have only node and lookback
+	if canAttemptPushdown(metadata) {
+		agg, node, err := tryPushDown(metadata)
+		if err == nil && agg != nil {
+			return agg, node, nil
+		}
 	}
 
-	//switch on the current node being processed
-	vs, isVectorSelector := qh.CurrentNode.(*parser.VectorSelector)
+	defaultAggregators := &aggregators{
+		timeClause:  "array_agg(time)",
+		valueClause: "array_agg(value)",
+		unOrdered:   false,
+	}
+
+	return defaultAggregators, nil, nil
+}
+
+func canAttemptPushdown(metadata *promqlMetadata) bool {
+	path := metadata.path // PromQL AST.
+	queryHints := metadata.queryHints
+	selectHints := metadata.selectHints
+	return extension.ExtensionIsInstalled && queryHints != nil && !hasSubquery(path) && selectHints != nil
+}
+
+// tryPushDown inspects the AST above the current node to determine if it's
+// possible to make use of a known pushdown function.
+// The following promQL functions can be pushed down:
+//   `delta`, `increase`, `rate`
```

Comment:
While this list is nice, it can get out of date pretty soon. Maybe we can add a TODO to reference a structure where these will be stored in the code?

Reply:
Maybe refer to `tryExtractPushdownableFunctionName`.

Reply:
👍